### PR TITLE
Add weight decay to SGD

### PR DESCRIPTION
**Pull request by cBournhonesque**
PR to add weight_decay to SGD.

Weight_decay is added after any momentum updates, so is NOT equivalent to L2 regularization.
(this is the similar to what is done in [tensorflow](https://github.com/tensorflow/addons/blob/v0.17.0/tensorflow_addons/optimizers/weight_decay_optimizers.py#L53) or AdamW in Pytorch).

Note: I had to add `HasArrayData` to `GradientProvider` because weight_decay makes use of the array data; I hope that's ok. Is there a case where we want to distinguish between a GradientProvider that can access the array data and one that cannot?

Related issue: #99

### CODE REVIEW

File: src/optim/sgd.rs [hunk 1]
```diff
@@ -244,6 +276,54 @@ mod tests {
...
+            [0.9988, 0.997, 0.995, 0.979, 0.799],
+            [0.99760115, 0.994003, 0.990005, 0.958021, 0.59820104],
+            [0.9964036, 0.991009, 0.98501503, 0.937063, 0.39760286],
+            [0.9952072, 0.988018, 0.98003, 0.9161259, 0.19720526],
+            [0.994012, 0.98502994, 0.97505, 0.8952098, -0.0029919297],
+        ];
+        for e in expected.iter() {
+            let gradients = backward((t.trace() * &rate).mean());
+            sgd.update(&mut t, gradients).expect("");
+            assert_eq!(t.data(), e);
```
Comment (coreylowman) [inline src/optim/sgd.rs:299]: For non-exact results you can use `crate::tests::assert_close` instead of `assert_eq!`. Have been slowly moving tests to use this instead, but all the optimizers still use assert_eq

**Comment by coreylowman** [comment]
Added #260 for the HasArrayData, that can probably be a PR separate from this for easy merging.

File: src/optim/sgd.rs (hunk 1 above)
Comment (coreylowman) [reply src/optim/sgd.rs:299]: Added #259 for this

File: src/optim/sgd.rs [hunk 2]
```diff
@@ -240,7 +297,111 @@ mod tests {
...
+            assert_close(t.data(), e);
+        }
+
+        // Should be equivalent to l2 regularization, even with momentum
+        t = Tensor1D::ones();
+        for e in expected.iter() {
+            let normal_loss = (t.trace() * &rate).mean();
+            let (normal_loss, tape) = normal_loss.split_tape();
+            let l2_loss = mul_scalar(t.clone().put_tape(tape).powi(2).sum(), weight_decay / (2.0));
+            let loss = add(l2_loss, &normal_loss);
```
Comment (coreylowman) [inline src/optim/sgd.rs:400]: Nice 😁

**Review by coreylowman — APPROVED** [review]
Nice implementation and great unit tests, thanks for this!
